modules/service/src/main/scala/lucuma/odb/service/UserInvitationService.scala

```diff
@@ -0,0 +1,212 @@
+
+          // Science users can only create CoI or Observer invitations, and only if they're the PI
+          case StandardRole.Pi(_)     => 
+            input match
+              case CreateUserInvitationInput.Coi(pid)      => createPiInvitation(pid, ProgramUserRole.Coi)
+              case CreateUserInvitationInput.Observer(pid) => createPiInvitation(pid, ProgramUserRole.Observer)
+              case _                                       => Result.failure("Science users can only create co-investigator and observer invitations.").pure[F]
+
+      def redeemUserInvitation(input: RedeemUserInvitationInput)(using Transaction[F]): F[Result[UserInvitation.Id]] =
+        user match
+          case GuestUser(_)                      => Result.failure("Guest users cannot redeemed user invitations.").pure[F]
+          case ServiceUser(_, _)                 => Result.failure("Service users cannot redeemed user invitations.").pure[F]
+          case StandardUser(_, _, _, c_duration) =>                  
+            val status = if input.accept then UserInvitation.Status.Redeemed else UserInvitation.Status.Declined
+            session
+              .prepareR(Statements.redeemUserInvitation)
+              .use(_.option(user, status, input.key))
+              .flatMap:
+                case None => Result.failure("Invitation is invalid, or has already been accepted, declined, or revoked.").pure[F]
+                case Some(r, ot, op, pid) =>
+                  val xa = transaction
+                  xa.savepoint.flatMap: sp =>
+                    session
+                      .prepareR(ProgramService.Statements.LinkUser.command)
+                      .use(_.execute(pid, user.id, r, ot, op))
+                      .as(Result(input.key.id))
+                      .recoverWith:
+                        case SqlState.UniqueViolation(_) => 
+                          xa.rollback(sp).as:
+                            Result.warning("You are already in the specified role; no action taken.", input.key.id)
+                    
+      def revokeUserInvitation(input: RevokeUserInvitationInput)(using Transaction[F]): F[Result[UserInvitation.Id]] =
+        user.role.access match
+          case Access.Guest => Result.failure("Guest users cannot revoke invitations.").pure[F]
+          case Access.Admin | Access.Service | Access.Staff =>
+            session.prepareR(Statements.revokeUserInvitationUnconditionially).use: pq =>
+              pq.option(input.id).map:
+                case Some(id) => Result(id)
+                case None     => Result.failure(s"Invitation does not exist or is no longer pending.")
+          case Access.Ngo | Access.Pi =>
+            session.prepareR(Statements.revokeUserInvitation).use: pq =>
+              pq.option(input.id, user.id).map:
+                case Some(id) => Result(id)
+                case None     => Result.failure(s"Invitation does not exist, is no longer pending, or was issued by someone else.")
```

Comment:
Grackle should maybe have some `Option` to `Result` `.toSuccess` thing like the `.toRight`.

Reply:
Never mind, I guess that is there in `grackle.syntax`.  It wouldn't really make any difference to employ it though.